contributors/design-proposals/api-machinery/customresources-versioning.md

```diff
@@ -0,0 +1,177 @@
+
+```golang
+// CustomResourceDefinitionSpec describes how a user wants their resource to appear
+type CustomResourceDefinitionSpec struct {
+  // Group is the group this resource belongs in
+  Group string
+  // Version is the version this resource belongs in
+  // must be always the first item in Versions field if provided.
+  Version string
+  // Names are the names used to describe this custom resource
+  Names CustomResourceDefinitionNames
+  // Scope indicates whether this resource is cluster or namespace scoped.  Default is namespaced
+  Scope ResourceScope
+  // Validation describes the validation methods for CustomResources
+  Validation *CustomResourceValidation
+
+  // ***************
+  // ** New Field **
+  // ***************
+  // Versions is the list of all supported versions for this resource.
+  // Validation: All versions must use the same validation schema for now. i.e., top 
+  // level Validation field is applied to all of these versions.
+  // Order: The order of these versions is used to determine the order in discovery API
+  // (preferred version first).
+  // The versions in this list may not be removed if they are in 
+  // CustomResourceDefinitionStatus.StoredVersions list.
+  Versions []CustomResourceDefinitionVersion
+}
+
+// ***************
+// ** New Type **
+// ***************
+type CustomResourceDefinitionVersion {
+  // Name is the version name, e.g. "v1", “v2beta1”, etc.
+  Name string
+  // Served is a flag enabling/disabling this version from being served via REST APIs
+  Served Boolean
+  // Storage flags the release as a storage version. There can be only one version
+  // flagged as Storage.
+  Storage Boolean
+}
+```
+
+The Status object will have a list of potential stored objects. This data is necessary to do a storage migration in future (the author can choose to do the migration themselves but there is a plan [TODO(lavalamp): link?] to solve the problem of migration, potentially for both standard and custom types).
+
+```golang
+// CustomResourceDefinitionStatus indicates the state of the CustomResourceDefinition
+type CustomResourceDefinitionStatus struct {
+  ...
+
+  // StoredVersions are all versions ever marked as storage in spec. Tracking these
+  // versions allow a migration path for stored version in etcd. The field is mutable
+  // so the migration controller can first make sure a version is certified (i.e. all
+  // stored objects is that version) then remove the rest of the versions from this list.
+  // None of the versions in this list can be removed from the spec.Versions field.
+  StoredVersions []string
+}
+```
+
+Basic validations needed for the `version` field are:
+
+* `Spec.Version` field exists in `Spec.Versions` field (serve should be True for that version if there is any).
```

Comment:
so we force at least one served version. Do we want that? Do we even have to (to allow e.g. the GC controller to do its work)?

Reply:
I think we don't want that. We discussed this on the implementation PR. I will amend this.